integrations/opensearch/src/haystack_integrations/components/retrievers/open_search_hybrid_retriever.py

```diff
@@ -0,0 +1,322 @@
+                "filters": self.filters_embedding,
+                "top_k": self.top_k_embedding,
+                "filter_policy": self.filter_policy_embedding,
+                "custom_query": self.custom_query_embedding,
+            },
+            "document_joiner": {
+                "join_mode": self.join_mode,
+                "weights": self.weights,
+                "top_k": self.top_k,
+                "sort_by_score": self.sort_by_score,
+            },
+        }
+
+        # look for extra kwargs for each component and add the document store as init param for the retrievers
+        if "text_embedder" in kwargs:
+            init_args["text_embedder"].update(kwargs["text_embedder"])
+        if "bm25_retriever" in kwargs:
+            init_args["bm25_retriever"].update(kwargs["bm25_retriever"])
+            init_args["bm25_retriever"]["document_store"] = self.document_store
+        if "embedding_retriever" in kwargs:
+            init_args["embedding_retriever"].update(kwargs["embedding_retriever"])
+            init_args["embedding_retriever"]["document_store"] = self.document_store
+
+        self.pipeline = self._create_pipeline(init_args)
+
+    def _create_pipeline(self, data: dict[str, Any]) -> Pipeline:
+        """
+        Create the pipeline for the OpenSearchHybridRetriever.
+
+        If the template for the ChatPromptBuilder is not provided, a default template is used.
+        """
+        text_embedder = SentenceTransformersTextEmbedder(**data["text_embedder"])
+        embedding_retriever = OpenSearchEmbeddingRetriever(**data["embedding_retriever"])
+        bm25_retriever = OpenSearchBM25Retriever(**data["bm25_retriever"])
+        document_joiner = DocumentJoiner(**data["document_joiner"])
+
+        hybrid_retrieval = Pipeline()
+        hybrid_retrieval.add_component("text_embedder", text_embedder)
+        hybrid_retrieval.add_component("embedding_retriever", embedding_retriever)
+        hybrid_retrieval.add_component("bm25_retriever", bm25_retriever)
+        hybrid_retrieval.add_component("document_joiner", document_joiner)
+
+        hybrid_retrieval.connect("text_embedder", "embedding_retriever")
+        hybrid_retrieval.connect("bm25_retriever", "document_joiner")
+        hybrid_retrieval.connect("embedding_retriever", "document_joiner")
+
+        # Define how pipeline inputs/outputs map to subcomponent inputs/outputs
+        self.input_mapping = {
+            # The pipeline input "query" feeds into each of the retrievers
+            "query": ["text_embedder.text", "bm25_retriever.query"],
+        }
+        # The pipeline output "answers" comes from "answer_builder.answers"
+        self.output_mapping = {"document_joiner.documents": "documents"}
+
+        return hybrid_retrieval
+
+    def to_dict(self):
+        """
+        Serialize OpenSearchHybridRetriever to a dictionary.
+
+        :returns:
+            Dictionary with serialized data.
+        """
+        return default_to_dict(
```

Comment:
We're not serializing the extra kwargs right now. I wonder if we could do that. For example by listing all parameters explicitly with the default that is used in the respective definitions of the components.
The `token` parameter of the `SentenceTransformersTextEmbedder` is a good example where serialization is not trivial but currently requires `token=self.token.to_dict() if self.token else None`

Reply:
> For example by listing all parameters explicitly with the default that is used in the respective definitions of the components

I'm a bit confused by this - the idea of the extra kwargs is to allow the user to customise other parameters that the super component doesn't expose, or missed what you were trying to suggest.

I have a solution to serialise the extra kwargs and then use them in the initialisation. I'm writing some tests now to make sure nothing breaks.

I think the `token` is a good example of a parameter that should be exposed, due to not being a regular parameter in serialisation/deserialisation, not having in the kwargs and as an explicit parameter might make things easier.